src/s3_plugin.js

```diff
@@ -93,19 +96,27 @@ module.exports = class S3Plugin {
       }
 
       if (isDirectoryUpload) {
-        let dPath = this.addSeperatorToPath(this.options.directory)
-        this.getAllFilesRecursive(dPath)
-          .then(this.filterAndTranslatePathFromFiles(dPath))
-          .then(this.filterAllowedFiles.bind(this))
-          .then(this.uploadFiles.bind(this))
-          .then(this.changeHtmlUrls.bind(this))
-          .then(this.invalidateCloudfront.bind(this))
-          .then(() => cb())
-          .catch(e => {
-            compilation.errors.push(new Error(`S3Plugin: ${e}`))
-            cb()
-          })
-      } else {
+        this.addGitHashToBasePath(this.options.basePath)
```

Comment:
I don't like the way this works, by having the promises inside of a callback of the `.then`. It would be better to do something like this 
```
this.setBasePath(this.options.basePath)
  .then(() => this.getAllFilesRecursive(dPath))
  .then(this.filterAndTranslatePathFromFiles(dPath))
 ... rest of .thens
```